### ISSUE TITLE
Threads support

### ISSUE THREAD

**Issue by laem** [element-hq/element-x-android#2831]
### Your use case

Hi, I'm missing the threads in ElementX. 

When some people in a room use threads, reading the messages in elementX becomes a mess. 

It's also a vert useful feature even on simple friends messaging groups. 

Will it be implemented soon ?

Regards

### Are you willing to provide a PR?

Yes

**Comment by codedust** [comment]
An interim solution would be to provide the option to create a thread via the context menu of a message.

Currently, the context menu of messages inside a thread is as follows:

![grafik](https://github.com/user-attachments/assets/17a78d8f-a73c-4f21-890c-74ade77e7e86)

The context menu of regular messages looks like this:

![grafik](https://github.com/user-attachments/assets/c97a0750-999f-4123-967c-916a29ae0cf1)

One could add the option "Reply in thread" here, too.

**Comment by boom-bang** [comment]
This is still an issue why we don't move the entire team to Element X from a legacy client

Threads are must have in 2k24

**Comment by domestomas** [comment]
Could maybe someone from the dev team give some info about when we can expect full threads support? I was not able to find any roadmap or any other kind of information about it. For me this (and spaces) is the top missing feature in element-X. It would be also nice to know the motivation behind prioritising other features like image captions, gallery and similar. I do not mean to sound ungrateful, I realy appreciate the work you are doing on element, It would by just nice to know what and why is going on. Maybe you could even publish some broader roadmap for 2025 as a christmas gift for all fans of Element and Matrix(?)

**Comment by manuroe** [comment]
Element X development still depends on features funding. The features you listed have been sponsorised. They are quite useful to make EX closer to mainstream messaging apps for non techy users. We basically follow what Element's customers need and pay so that we can continue to work on the project. 

We also consider threads and spaces to be super important too. We already worked a lot on them on the product and UX sides with design prototypes, user and customer interviews., etc. This work is not visible here on GH but it takes a huge amount of time. I can confidently say those features will be more useful (even usable) for end users when we will land them in the EX apps. Next step for us is to align stars (and sponsorship) to priorisite the implementation. Hopefully, it will happen soon.

As you may know, funding a FOSS project can be a rollercoaster. We stopped publishing our roadmap long time ago because people did not want to sponsor a feature Element is going to build for free for them.

**Comment by LecrisUT** [comment]
Why not show in the roadmap unsponsored features (in place of backlog), along with the priority of the sponsored ones.

**Comment by ajkessel** [comment]
Have you considered implementing some band-aid approaches in the meantime? When you reply to a message in a thread in gomuks, you have the option to have it "add to thread" or "reply to message in thread". Maybe you don't want temporary/half solutions, but if it's going to be months or years before threads exist, maybe something easier to implement would make sense in the meantime, at least as a labs feature?

**Comment by Midar** [comment]
+1, the lack of threads and Spaces is a hard blocker for many people to migrate. If Element wants people to move to Element X, these are requirements

I really wish Element wouldn’t try to push Element X until it’s ready. Right now we have two incompatible clients, especially when it comes to phone calls.

**Comment by boom-bang** [comment]
I’m going to be blunt: the current stance on prioritizing Threads is completely off-track. Element X’s product management approach here is, frankly, 100% wrong. When you’re trying to attract serious enterprise users, especially the thousands of organizations currently relying on Slack, Threads are not an optional “nice-to-have.” They are the core functionality many businesses require before even considering a migration.

Here’s the evidence:

1. **Established Precedent in Enterprise Tools**
Slack, Microsoft Teams, Google Chat, all of these platforms have robust threaded conversations. This isn’t a secondary feature; it’s central to how large teams reduce noise and maintain clarity. If Element X doesn’t match or improve upon this standard, it won’t even appear on the radar of companies looking to switch.

2. **Essential for Large-Scale Collaboration**
Threads let teams break off side-discussions without cluttering the main conversation. When you have hundreds or thousands of employees in a room, unthreaded chats become chaotic and unmanageable. Decision-making slows to a crawl as people scramble to keep track of who said what and when.

Threads drastically improve signal-to-noise ratio. They keep tangential discussions, clarifications, and question/answer pairs neatly organized, so the main channel can stay focused on the central topic.

3. **Critical for Onboarding and Retaining Users**
Migrating from Slack to Element X is not just about replicating chat logs; it’s about preserving familiar workflows so teams don’t skip a beat. Many Slack power-users rely on threads every day for project management, quick feedback loops, and knowledge-sharing.

Without fully functional threading, any attempt to move over from Slack fails instantly because user friction shoots up. Confusion, frustration, and churn will follow, people simply won’t adopt a solution missing the single largest advantage of Slack-like platforms.

4. **Evidence from User Feedback**
The conversation around “Threads support” is consistently the most common (and often heated) request from potential enterprise adopters. Dozens of open-source communities, software developers, and corporate teams have all said: “We want to move to Matrix/Element, but we can’t without Threads.”

Every time the topic comes up, you see the same points repeated: there are “two incompatible clients,” confusion around half-baked or incomplete features, and a total roadblock for any serious Slack replacement strategy.


5. **Long-Term Viability and Growth**
If the plan is to position Element X as a true Slack or Teams competitor, ignoring the importance of Threads will only repel exactly the customers who might be paying for these enterprise features.

Threads are not just a “techy user” feature. They are a basic organizational tool that influences how entire departments communicate. Any sales pitch or demonstration of Element X to large companies or governments will fall flat without it.

Element X can’t fully claim _feature parity_ with other mainstream chat tools if Threads remain an afterthought. The entire premise of a Slack alternative hinges on the ability to handle focused, parallel discussions at scale, and that’s exactly what threads enable.

In short, Thread support must be top priority. It is the gatekeeper for large-scale adoption, far more critical than nice-to-have additions like image captions or gallery views. Without robust threading, thousands of interested organizations (my team included) simply cannot move from Slack to Element X. If the goal is to grow and thrive in this space, Threads aren’t optional; they’re fundamental.

**Comment by leonardehrenfried** [comment]
> Element X development still depends on features funding. The features you listed have been sponsorised. They are quite useful to make EX closer to mainstream messaging apps for non techy users. We basically follow what Element's customers need and pay so that we can continue to work on the project.
> 
> We also consider threads and spaces to be super important too. We already worked a lot on them on the product and UX sides with design prototypes, user and customer interviews., etc. This work is not visible here on GH but it takes a huge amount of time. I can confidently say those features will be more useful (even usable) for end users when we will land them in the EX apps. Next step for us is to align stars (and sponsorship) to priorisite the implementation. Hopefully, it will happen soon.
> 
> As you may know, funding a FOSS project can be a rollercoaster. We stopped publishing our roadmap long time ago because people did not want to sponsor a feature Element is going to build for free for them.

**Comment by boom-bang** [comment]
@leonardehrenfried we need a paradigm shift here. The conversation isn’t actually about money or funding. That mindset misses the fundamental point: success hinges on delivering the features and experience people need, not on how much budget you can throw at the problem. Slack and Teams didn’t blow up just because they had big funding, they solved a real workflow issue by making threads the heart of team collaboration. If Element X wants that same level of impact, it has to prioritize what users truly care about. Budgets and funding follow when the product delivers genuine value. Shift the focus from finances to the actual user experience, and the rest will fall into place.

**Comment by leonardehrenfried** [comment]
Ahhh, I see. If only they cared about user experience...

**Comment by Midar** [comment]
If all Element X cares about is getting money to get features funded and nothing else, that's fine. But then please, stop trying to push it as a replacement for Element, because it clearly isn't. And not only is it not, you clearly don't want it to be a replacement, as it's not even planning to be feature complete to not risk funding. At this point, the damage to the ecosystem is immense. Please stop pushing Element X and instead tell people to go back to Element.

**Comment by laem** [comment]
Element X is currently only usable for conversations with friends and small groups, like Whatsapp. 
It's getting pushed for this usage. Element is totally usable for work-related usages.

**Comment by pangyuehung** [comment]
As [mentioned here](https://github.com/element-hq/element-meta/issues/1915#issuecomment-2672137056), does this ticket actually belong against https://github.com/element-hq/element-meta instead? Also, is there a ticket for spaces support somewhere?

**Comment by filipesmedeiros** [comment]
Ok… that's big ahah

But then a question remains: I suppose these threads 2.0 are going to be backwards-INcompatible with the current threads? Is this something that the Matrix community really wants to do? Has this been discussed publicly? Or only internally?

**Comment by manuroe** [comment]
Do not be scared. 2.0 is more a marketing version. There will be no incompatibility issue in the Matrix ecosystem.
Threads 1.0 is what the protocol currently defines. Threads 2.0 will use Threads 1.0 and will add this participation model in a new separated API. 

Threads 2.0 clients will be able to communicate with Threads 1.0 clients and vice versa.
The difference will be on notifications: Threads 2.0 clients users will get notifications for messages in a thread while it will stay "complicated" for Threads 1.0 clients users

**Comment by estux** [comment]
> Several X in $X00k. We learnt that threads as they are specified in the matrix protocol are not enough for a good UX. We want to add a participation model at the top of them. With threads 2.0, users will be automatically notified for new messages in threads they participated in. There will be a mechanism to register and unregister notifications for a threads. If I can make a comparison with GH, it is going to be very similar as how you have notifications for GH issues.
> 
> Threads 2.0 is to be defined. We need a new MSC for this new Client-Server API and implementation in Synapse, EX and EW.

Please consider to add also a way to tag the threads as "closed/completed" or "open" and a way to hide/archive threads. These solutions would be very useful to periodically clear the clutter and maintain a clear vision of stuff going on. Thank you!

**Comment by heyakyra** [comment]
Has Threads 2.0 been planned with consideration towards [Zulip-style or Twist-style topic-based (threaded-only mode)](https://zulip.com/why-zulip/#zulips-unique-topic-based-threading-model-makes-efficient-communication-possible-heres-how) on a [per-room basis](https://github.com/element-hq/element-web/issues/2349#issuecomment-364887293) (or per-server)? 

From @KB1RD's [comment on the completed element-web ticket](https://github.com/element-hq/element-web/issues/2349#issuecomment-554011591), it sounded like https://github.com/matrix-org/matrix-spec-proposals/pull/2326 was work towards that use case, but maybe that has been abandoned since the v1 threading implementation landed.

**Comment by Ludwig4004** [comment]
Thank you for keeping this on the roadmap – I really appreciate the transparent communication and steady development!
Thread support is such an important feature, especially for active rooms where things can get messy without proper structure.
I’m genuinely looking forward to seeing this in Element X – thanks for all your hard work on it!

**Comment by laem** [comment]
Just tried threads on Element X after activating the experimental option, it works !

Great, no need to use Element classic anymore.

**Comment by leonardehrenfried** [comment]
OMG, thank you for this message! I have been waiting for this for a long time.

It works great - thanks to everyone involved!

**Comment by xpac1985** [comment]
Thanks for the hint. I just tried that, and now I'm... confused.
I get a "reply in thread" option for messages, yay, I no longer have to use a second client to start a thread or ask someone else to create one for me... that's great and was missing so much.
But - the same option also turns on typical threading under the initial message, while also removing the Thread messages from the general channel message flow - the only thing ElementX did better than the old client?
Now I have no idea when somebody writes a message in an old thread, and I would have to continously scroll back the last I-dunno-how-many pages to check all threads manually for new messages?

Am I missing something, or has this improvement really at the same time made Threads more usable and a lot less usable?

constructive criticism:
Can we please get two switches in lab settings - one two allow threaded reply on a non-thread message (for which the UI could simply be used as before when you replied to a threaded message), and one to enable that old Threads view, where you need to manually check all old threads, if somebody really prefers that option...

**Comment by ajkessel** [comment]
Wouldn't an "unread threads" view option be the best solution?

**Comment by xpac1985** [comment]
Personally, I prefer the inline messages in the channel over a separate view, because I usually read all threads in the channels I'm in, with very few exceptions, and considering how "well" the read status of the thread view in Desktop Element works, I'd also prefer to stick with the old version - I'm really just looking for the "reply in thread" button for messages that are not yet threaded.
I can understand how others would prefer a thread overview, and think both are legitimate options, but currently I'm stuck between two choices - not being able to start a thread at all, or missing most messages sent in threads that are not the most recent one. If that's the choice, I prefer option A, and have to start the old Element for the few cases where it needs to be me who starts a thread.

(And honestly I'd so much prefer if every chat app would take over Zulip's approach to threading, but I guess that will never happen :))

**Comment by mamo2021** [comment]
Hello everyone,

we work as a team here and we also miss the threads view as it was in the old Element. Threads should be displayed on mobile phones and similar devices just as they are on the desktop — that is, with a thread overview and the messages for each thread. You’ll never please everyone, but displaying thread messages inline without the option to read the thread without the unrelated messages is a bad solution.

**Comment by jmartinesp** [comment]
Threads support can be enabled in `Settings -> Labs -> Enable thread replies`, it's not enabled by default yet because it lacks some support from the Rust SDK yet, but it should be good enough for daily use.

**Comment by mamo2021** [comment]
You're right. I didn't knew this. Sorry.
Thanks.

**Comment by boom-bang** [comment]
Wow, impressive, let's check 🚀

**Comment by bxdxnn** [comment]
@manuroe

> As you may know, funding a FOSS project can be a rollercoaster. We stopped publishing our roadmap long time ago because people did not want to sponsor a feature Element is going to build for free for them.

Sorry, but I don't see anywhere how a person could sponsor features or at very least donate to Element?
If by "funding" means buying an Enterprise version, that's a no-go for the most people as they won't be able to buy it at all. Element Pro is a product specifically for organizations/businesses, not ordinary people.

**Comment by Susurrus** [comment]
So threads has been supported for a while, so I think this can be closed.  I think some other features around threading have been mentioned here, but those should be raised as new issues instead if they're still relevant.

**Comment by meyerrj** [comment]
One key feature that is still missing for me is the ability to access a list of threads like you can on the element web/desktop client. 

Having to scroll up through conversation history to hunt down a thread is a rather frustrating experience.  

If I could at least just tap the room name at the top, and see "Threads" as a menu option  and then select it from there, that would be a significant improvement to what we currently have.

**Comment by Susurrus** [comment]
That exists. In rooms with threads, there's the threads button, just like on Web. (See Screenshot).

<img width="1080" height="2399" alt="Image" src="https://github.com/user-attachments/assets/96a8809d-8c57-4d72-b6a8-09a471a26b84" />

But again, the original issue by @laem here is addressed: threads are supported. If you want additional functionality around threads, please first confirm it doesn't already exists, and check for an open issue about it first, and then open a new issue if not. Issues become a bit unmanageable if too much feature creep happens.

**Comment by Susurrus** [comment]
I just realized there's a "thread replies" feature in Labs, so you might need to turn that on to get the threads icon to show.

**Comment by mamo2021** [comment]
I tried this out for a while. It actually would make sense to have a thread overview, especially for threads that go on for a long time, when scrolling becomes tedious. The thread stays at the initial post in the timeline—that is, at the farthest point. Meanwhile, if the conversation moves along quickly, you always have to scroll a lot.
It would also be useful if you could assign topic headings to threads, which would then be visible in this overview or list that still needs to be created. That way, an ongoing conversation could be categorized by topic quite easily.
By the way, I can’t see the “Threads” button on my end. It doesn’t appear, even though there are clearly threads.

**Comment by meyerrj** [comment]
@Susurrus I have "threads replies" turned on in labs, using version 26.09.1 (202609010). 

I do not see that icon on my mobile chats, but I do in desktop.  Perhaps it only shows when there's unread activity in my threads?  If that's the case, I think that is a usability issue that should be addressed.

Edit: Nope, found "unread" threads in some conversations on desktop, and I still don't see the threads button on mobile.

**Comment by bxdxnn** [comment]
> I do not see that icon on my mobile chats

Tap 7 times on the app version in settings and enable threads list in the developer options.

**Comment by meyerrj** [comment]
Thanks, that seems to have done the trick for me. 

Maybe these hidden threads options should be rolled into labs, or made more available under Advanced settings if you enable the threads options in labs?

**Comment by bxdxnn** [comment]
> Maybe these hidden threads options should be rolled into labs, or made more available under Advanced settings if you enable the threads options in labs?

I agree, but no response from devs about that, feels like Threads aren't a priroity right now.

**Comment by xpac1985** [comment]
I'll try not be an ass here, but considering "Threads support done", when 80% of basic Thread support is either still in Labs or hidden behind some developer options magic, and I still cannot even start a thread in the app - just reply to it once someone else started it first?

I *really* value the effort people have put into this, and I have at least a rough idea how much work a lot of this is.
And, to be fair, the initial feature request was sparse, to be nice.

So - after venting without trying to vent too much :D - let me try to put my understanding of "basic threads support" into a few lines:
- Being able to start a new thread somehow. A "reply in thread" in the context menu of an existing message would be enough, I guess that should be easy enough, as there's already support to reply to threaded messages.
- Being able to read threaded messages, in general . This one is in.
- Being able to get a list of existing threads - else it get's really confusing really fast
- Give me a chance to decide if I wanna see all threaded messages inline in the regular room view, or only the starting one, and an indicator for thread activity somewhere.
- This is already better in the mobile app than on the Desktop - because inline thread messages exist. Until I read about the "thread list" option in developer settings, turning on the Lab Threads thing gave me a thread list, but removed the inline view. So this is actually better (for me)

So I'd say - threads support needs at least an option to actually start a thread, and for the features to be at least accessible from Labs, or better "fully in" ;-)

**Comment by bxdxnn** [comment]
@xpac1985 everything you've listed is already possible once you enable it in the Labs and that feature flag. The Thread support is locked behind a flag mostly because there are still some improvements to be done, e.g. the most blocking ones would be that I can't see if there are unread messages/mentions in a Thread, sometimes I don't see new messages in a Thread, etc.

CC @mxandreas

**Comment by xpac1985** [comment]
So - when I enable Threads in Labs - I don't have a choice to still see new thread messages inline in the channel - I'd have to always look at the threads list, and manually check each thread for new messages, right? Or is there another hidden setting that I missed to get enabled again after enabling Threads in Labs?
